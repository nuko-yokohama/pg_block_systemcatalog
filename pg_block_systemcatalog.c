/*------------------------------------------------------------------------------
 * pg_block_systemcatalog.c
 *
 * PostgreSQL extension to block references to system catalogs.
 *
 * author @nuko_yokohama
 *
 *------------------------------------------------------------------------------
 */
#include "postgres.h"

#include "executor/executor.h"
#include "executor/spi.h"
#include "miscadmin.h"
#include "libpq/auth.h"
#include "tcop/utility.h"
#include "utils/builtins.h"
#include "utils/guc.h"
#include "utils/memutils.h"
#include "utils/timestamp.h"
#include "libpq/auth.h"
#include "port.h"
#include "catalog/catalog.h"


#include "pgtime.h"
#include <time.h>

PG_MODULE_MAGIC;

extern bool superuser();

/*
 * GUC variable fot pg_block_systemcatalog
 *
 * Role name to allow a reference to the system catalog.
 *
 */
char *allow_role = NULL;

void _PG_init(void);


/*
 * Hook functions
 */
static ExecutorCheckPerms_hook_type next_ExecutorCheckPerms_hook = NULL;

static bool checkRole(void)
{
    Oid allow_roleOid;

    if (superuser())
        return true;

    if (allow_role == NULL)
        return false;

    allow_roleOid = get_role_oid(allow_role, true);
    ereport(DEBUG1, (errmsg("allow_role = %s, allow_roleOid = %d, GetUserId() = %d.", 
        allow_role, allow_roleOid, GetUserId())));

    /*
     * Is the current user included in members of allow_role?
     */
    if (is_member_of_role(allow_roleOid, GetUserId()))
        return true;

    return false;
}

static void checkTables(List * rangeTabls)
{
    ListCell *lr;
    foreach(lr, rangeTabls)
    {
        RangeTblEntry *rte = lfirst(lr);
        Relation rel;
        Oid relOid;
        bool allow = true;

        /*
         * 1: superuser or allow_role (Roles can refer to the system catalog.)
         * 0: other role/user (Roles can not refer to the system catalog.)
         */
        allow = checkRole();

        if (!allow) {
            /*
             * To check whether the query to perform the access to the system catalog.
             */
            relOid = rte->relid;
            rel = relation_open(relOid, NoLock);

            if (IsSystemNamespace(RelationGetNamespace(rel)))
            {
                relation_close(rel, NoLock);
                ereport(ERROR, (errcode(ERRCODE_INVALID_SCHEMA_DEFINITION),
                    errmsg("pg_block_systemcatalog: Reference to the system catalog is not permitted.")));
            }
            relation_close(rel, NoLock);
        }
    } 
}

/*
 * Hook ExecutorCheckPerms_hook.
 */
static bool
pg_bsc_ExecutorCheckPerms_hook(List *rangeTabls, bool abort)
{

    checkTables(rangeTabls);

    /* Call the next hook function */
    if (next_ExecutorCheckPerms_hook &&
        !(*next_ExecutorCheckPerms_hook) (rangeTabls, abort))
        return false;

    return true;
}


void
_PG_init(void)
{
    /* Must be loaded with shared_preload_libaries */
    if (IsUnderPostmaster)
        ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
                errmsg("pg_block_systemcatalog must be loaded via shared_preload_libraries")));

    /*
     * pg_block_systemcatalog.allow_role
     */
    DefineCustomStringVariable(
        "pg_block_systemcatalog.allow_role",
        "Role name to allow a reference to the system catalog.",
        NULL,
        &allow_role,
        NULL,
        PGC_SUSET,
        GUC_NOT_IN_SAMPLE,
        NULL, NULL, NULL);

    /*
     * Install our hook functions after saving the existing pointers to
     * preserve the chains.
     */
    next_ExecutorCheckPerms_hook = ExecutorCheckPerms_hook;
    ExecutorCheckPerms_hook = pg_bsc_ExecutorCheckPerms_hook;

    /* Log that the extension has completed initialization */
    ereport(LOG, (errmsg("pg_block_systemcatalog.")));

}

