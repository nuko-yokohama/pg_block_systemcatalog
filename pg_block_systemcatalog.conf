# pg_block_systemcatalog configuration
shared_preload_libraries = pg_block_systemcatalog
