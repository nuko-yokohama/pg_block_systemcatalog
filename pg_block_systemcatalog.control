# pg_block_systemcatalog extension
comment = 'PostgreSQL extension to block references to system catalogs.'
default_version = '1.0'
module_pathname = '$libdir/pg_block_systemcatalog'
relocatable = true
