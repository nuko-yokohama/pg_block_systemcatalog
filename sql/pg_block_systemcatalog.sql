CREATE ROLE monitor_role;
CREATE USER monitor;
CREATE USER test;
GRANT monitor_role TO monitor;

CREATE TABLE tmp (id int, data text);

SET pg_block_systemcatalog.allow_role = '';

-- superuser
SELECT CURRENT_USER;

SELECT relname FROM pg_class WHERE relname = 'tmp';
SELECT table_name FROM information_schema.tables WHERE table_name = 'tmp';
SELECT relname FROM pg_stat_user_tables WHERE relname = 'tmp';

-- test(error)
SET ROLE test;
SELECT CURRENT_USER;

SELECT relname FROM pg_class WHERE relname = 'tmp';
SELECT table_name FROM information_schema.tables WHERE table_name = 'tmp';
SELECT relname FROM pg_stat_user_tables WHERE relname = 'tmp';

RESET ROLE;

-- monitor(error)
SET ROLE monitor;

SELECT CURRENT_USER;

SELECT relname FROM pg_class WHERE relname = 'tmp';
SELECT table_name FROM information_schema.tables WHERE table_name = 'tmp';
SELECT relname FROM pg_stat_user_tables WHERE relname = 'tmp';

RESET ROLE;

SET pg_block_systemcatalog.allow_role = 'monitor';

-- monitor(success)
SET ROLE monitor;

SELECT CURRENT_USER;

SELECT relname FROM pg_class WHERE relname = 'tmp';
SELECT table_name FROM information_schema.tables WHERE table_name = 'tmp';
SELECT relname FROM pg_stat_user_tables WHERE relname = 'tmp';

RESET ROLE;

DROP TABLE tmp;
REVOKE monitor_role FROM monitor; 
DROP ROLE test;
DROP ROLE monitor;
DROP ROLE monitor_role;
