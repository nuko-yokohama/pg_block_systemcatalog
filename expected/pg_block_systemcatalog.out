CREATE ROLE monitor_role;
CREATE USER monitor;
CREATE USER test;
GRANT monitor_role TO monitor;
CREATE TABLE tmp (id int, data text);
SET pg_block_systemcatalog.allow_role = '';
-- superuser
SELECT CURRENT_USER;
 current_user 
--------------
 postgres
(1 row)

SELECT relname FROM pg_class WHERE relname = 'tmp';
 relname 
---------
 tmp
(1 row)

SELECT table_name FROM information_schema.tables WHERE table_name = 'tmp';
 table_name 
------------
 tmp
(1 row)

SELECT relname FROM pg_stat_user_tables WHERE relname = 'tmp';
 relname 
---------
 tmp
(1 row)

-- test(error)
SET ROLE test;
SELECT CURRENT_USER;
 current_user 
--------------
 test
(1 row)

SELECT relname FROM pg_class WHERE relname = 'tmp';
ERROR:  pg_block_systemcatalog: Reference to the system catalog is not permitted.
SELECT table_name FROM information_schema.tables WHERE table_name = 'tmp';
ERROR:  pg_block_systemcatalog: Reference to the system catalog is not permitted.
SELECT relname FROM pg_stat_user_tables WHERE relname = 'tmp';
ERROR:  pg_block_systemcatalog: Reference to the system catalog is not permitted.
RESET ROLE;
-- monitor(error)
SET ROLE monitor;
SELECT CURRENT_USER;
 current_user 
--------------
 monitor
(1 row)

SELECT relname FROM pg_class WHERE relname = 'tmp';
ERROR:  pg_block_systemcatalog: Reference to the system catalog is not permitted.
SELECT table_name FROM information_schema.tables WHERE table_name = 'tmp';
ERROR:  pg_block_systemcatalog: Reference to the system catalog is not permitted.
SELECT relname FROM pg_stat_user_tables WHERE relname = 'tmp';
ERROR:  pg_block_systemcatalog: Reference to the system catalog is not permitted.
RESET ROLE;
SET pg_block_systemcatalog.allow_role = 'monitor';
-- monitor(success)
SET ROLE monitor;
SELECT CURRENT_USER;
 current_user 
--------------
 monitor
(1 row)

SELECT relname FROM pg_class WHERE relname = 'tmp';
 relname 
---------
 tmp
(1 row)

SELECT table_name FROM information_schema.tables WHERE table_name = 'tmp';
 table_name 
------------
(0 rows)

SELECT relname FROM pg_stat_user_tables WHERE relname = 'tmp';
 relname 
---------
 tmp
(1 row)

RESET ROLE;
DROP TABLE tmp;
REVOKE monitor_role FROM monitor; 
DROP ROLE test;
DROP ROLE monitor;
DROP ROLE monitor_role;
